presto-spi/src/main/java/io/prestosql/spi/connector/ProjectionApplicationResult.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.spi.connector;
+
+import io.prestosql.spi.expression.ConnectorExpression;
+import io.prestosql.spi.type.Type;
+
+import java.util.List;
+
+public class ProjectionApplicationResult<T>
+{
+    private final T handle;
+    private final List<ConnectorExpression> newProjections;
```

Comment:
What's the relation between `newProjections.get(i)` and `assignments.get(i)`?

Reply:
There's no relationship. `newProjections` is the list of projections that replaces the original projections. They might include references to old or new columns. `assignments` is the new set of columns exposed by the derived table described by `handle`.

Here's an example of how this is supposed to work:

Given this plan:

```
- project
        x = f1(a, b)
        y = f2(a, b)
        z = f3(a, b)
    - scan (TH0)
            a = CH0
            b = CH1
            c = CH2
```

The rule would call `applyProjection` with

```
handle = TH0
projections = [
    f1(a, b)
    f2(a, b)
    f3(a, b)
]
assignments = [
    a = CH0
    b = CH1
    c = CH2
]
```

Let's assume the connector only knows how to handle `f1(...)` and `f2(...)`. It would return:

```
handle = TH1
projections = [
    v2
    v3
    f3(v0, v1)
]
assignments = [
    v0 = CH0
    v1 = CH1
    v2 = CH3  (synthetic column for f1(CH0, CH1))
    v3 = CH4  (synthetic column for f2(CH0, CH1))
]
```

The rule then recreates the subplan as:

```
- project
        x = $2
        y = $3
        z = f3($0, $1)
    - scan(TH1)
            $0 = CH0
            $1 = CH1
            $2 = CH2
            $3 = CH3
```

Reply:
OK. Thanks for explanation. Should such explanation this be part of Javadoc?

Reply:
That’s a great idea. I’ll add it